cpu/msp430_common/cpu.c

```diff
@@ -14,23 +14,46 @@
 #include "thread.h"
 
 /*
- * we must prevent the compiler to generate a prologue or an epilogue
- * for thread_yield_higher(), since we rely on the RETI instruction at the end
- * of its execution, in the inlined __restore_context() sub-function
+ * This function can both be called from ISR and from thread context.
+ *
+ * In both cases, the caller will use "CALL", which pushes the return
+ * address on the stack before executing the function's first instruction.
+ *
+ * If called within ISR, just set sched_context_switch_request and
+ * directly return to the call site. A regular "ret" does this.
+ *
+ * If called from stack context, it needs to prepare the stack so it looks exactly
+ * as if the thread had been interrupted by an ISR, which requires the SR to be on
+ * stack right after the return address. So we do this manually.
+ * __save_context() will then save the remaining registers, then store the stack
+ * pointer in the thread's thread_t.
+ *
+ * At this point, the thread context is properly saved. sched_run (possibly) changes
+ * the currently active thread, which __restore_context() then restores, resuming
+ * execution at the call site using reti.
+ *
+ * The naked attribute is needed to ensure this works.
  */
 __attribute__((naked)) void thread_yield_higher(void)
 {
-    __asm__("push r2"); /* save SR */
-    irq_disable();
+    if (irq_is_in()) {
+        sched_context_switch_request = 1;
+        __asm__("ret");
```

Comment:
You should prefer [Extended Asm](https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html) over [Basic Asm](https://gcc.gnu.org/onlinedocs/gcc/Basic-Asm.html#Remarks) (more instances below)

Reply:
Citing from [How to Convert Basic asm to Extended asm](https://gcc.gnu.org/wiki/ConvertBasicAsmToExtended):

> gcc is considering changing the long-time semantics of basic asm. Instead of not clobbering anything, it may soon begin clobbering everything. This may fix subtle errors in existing code. However there is a (small) chance this will cause problems with existing, correctly function code. Further, this could also introduce performance issues as the contents of some/all registers (even ones your asm doesn't use) may need to be saved/reloaded around your asm statement. To "future-proof" your code against these types of changes, use extended asm and specify exactly what needs to be clobbered.

So assuming you don't want to touch this after a GCC update, you should use Extended Asm instead.

There is one pitfall: If an Extended Asm has no `volatile` keyword, it will be optimized out if the compiler deems it unused. Extended Asm without outputs has an implicit `volatile` keyword. So if there are outputs in Extended Asm, no explicit `volatile` keyword is given, and all outputs are deemed unused, the compiler optimized Extended Asm out. (I personally just always add `volatile` to ASM...)

Reply:
Please fix as follow up if you insist.
If I understand, the difference would be

basic:

    asm asm-qualifiers ( AssemblerInstructions )

vs extended:

asm asm-qualifiers ( AssemblerTemplate 
                 : OutputOperands 
                 [ : InputOperands
                 [ : Clobbers ] ])

As there are no output operands, input operands or clobbers to add, why bother? how would it look anyways?

Reply:
to keep current behavior, I'd need to add "volatile", and a ":" after the template? doesn't seem like an improvement.

Reply:
With basic ASM the semantics are not clear.

E.g. some compilers will spill *all* registers to RAM before each basic ASM statement, as the compiler cannot know which registers the basic assembly uses. This is a significant performance penalty.

Other compilers don't do anything, and if the assembly touches registers without backing them up and restoring them, it might override local variables. And as the register allocation can change with seemingly unrelated changes in the code or toolchain upgrades, this could be a pretty sneaky bug. (E.g. while testing a register is uses that did not hold a meaningful value, but later on the basic assembly breaks due to different register allocation.)

IMO there should be a CI test for basic ASM and Murdock should prevent merging it. Every Basic ASM statement can be converted into a safer and faster Extended ASM statement.

Reply:
This results in no change in ROM size:

```diff
diff --git a/cpu/msp430_common/cpu.c b/cpu/msp430_common/cpu.c
index 419e043d22..a69d5aa8c5 100644
--- a/cpu/msp430_common/cpu.c
+++ b/cpu/msp430_common/cpu.c
@@ -39,10 +39,14 @@ void thread_yield_higher(void)
         sched_context_switch_request = 1;
     }
     else {
-        __asm__("push r2"); /* save SR */
-        __asm__("dint");
-        __asm__("nop");
-
+        __asm__ volatile(
+            "push r2"               "\n\t" /* save SR */
+            "dint"                  "\n\t" /* reti will restore SR, and thus, IRQ state */
+            "nop"                   "\n\t" /* dint takes an additional CPU cycle to take into effect */
+            : /* no outputs */
+            : /* no inputs */
+            : /* no clobbers */
+        );
         __save_context();
 
         /* have sched_active_thread point to the next thread */
```

So apparently the current toolchain doesn't back up registers. As there are no data registers touched in the assembly, the basic assembly would be safe to use.

But note: Future versions of GCC might change the behavior of basic assembly. So with the Extended Asm, we don't need to worry about future regressions in ROM size and performance.

Reply:
applied.

> Every Basic ASM statement can be converted into a safer and faster Extended ASM statement.

that's just plain untrue. the different syntax does not result in different assembly, in this context. it is not safer or faster. and please don't claim you can make those three instructions safer and faster.

In this context, extended asm vs basic asm is pure cosmetics, and not for the better.

> But note: Future versions of GCC might change the behavior of basic assembly.

Yeah, well, there might be some backlash when all the RTOS break because gcc decided to change the semantics. These things do have some inertia, for a reason.

I guess this is a style matter.

Reply:
> it is not safer or faster.

On compiler that do back up all registers before every basic asm and restore them after every basic asm, Extended Asm is indeed faster. And on compilers that don't back up registers, a subtile bug is possible that only ever triggers when a different register allocation is chosen. Avoiding this possibility altogether is safer.

Thus, Extended Asm in general is safer and faster than Basic Asm.

> Yeah, well, there might be some backlash when all the RTOS break because gcc decided to change the semantics.

GCC developers are famous to care very little about backlash. If you can't convince them with reason, you won't convince them with backlash.

And they have explicitly warned (RTOS) developers to not use Basic ASM, as they plan to change the semantics in the future. IMO, we should take this warning seriously.